README.md
# gifencoder

[![Pub Package Version](https://img.shields.io/pub/v/gifencoder.svg)](https://pub.dartlang.org/packages/gifencoder)
[![Latest Dartdocs](https://img.shields.io/badge/dartdocs-latest-blue.svg)](https://pub.dartlang.org/documentation/gifencoder/latest)

A gif encoder written in Dart.

## Usage

To create a regular GIF:

```dart
import 'dart:html';
import 'package:gifencoder/gifencoder.dart';

int width = ...;
int height = ...;
var ctx = new CanvasElement(width: width, height: height).context2D;
// draw your image in the canvas context
var data = ctx.getImageData(0, 0, width, height);
List<int> bytes = gifencoder.makeGif(width, height, data.data)
```

To create an animated Gif, use a GifBuffer instead:

```dart
int framesPerSecond = ...;
var frames = new gifencoder.GifBuffer(width, height);
for (var i = 0; i < myFrameCount; i++) {
    // draw the next frame on the canvas context
    frames.add(ctx.getImageData(0, 0, width, height).data);
}
List<int> bytes = frames.build(framesPerSecond);
```

Once you have the bytes of the GIF, you can save it somewhere or convert it into a data URL. See `example/example.dart` for how to do that. You can run the example by running the following: `pub run build_runner serve` and then connecting to localhost in the browser on port 8081.

## Testing

To test in the VM:

```bash
pub run test -p vm
```

To run browser tests in Chrome:

```bash
pub run build_runner test --release -- -p chrome
```

You can also access the tests from a browser by running `pub run build_runner serve` and then connecting to localhost in the browser on port 8080.

CHANGELOG.md
# Changelog

## 1.0

- Supports and requires Dart 2 (no API change)

## 0.2.3

- Stable version for Dart 1


test/gif_test.dart
library gif_test;

import "package:gifencoder/gifencoder.dart" as gif;
import 'package:test/test.dart';

void main() {
  // Based on: http://www.matthewflickinger.com/lab/whatsinagif/bits_and_bytes.asp
  // Also see: http://en.wikipedia.org/wiki/File:Quilt_design_as_46x46_uncompressed_GIF.gif
  const headerBlock = const [0x47, 0x49, 0x46, 0x38, 0x39, 0x61]; // GIF 89a
  const oneByOneTwoColor = const [
    0x1, 0, 0x1, 0, 0xF0, 0, 0 // 1x1, 2 colors, background index 0
  ];
  const blackBlack = const [0, 0, 0, 0, 0, 0]; // 3 bytes per color
  const startOneByOneImage = const [0x2c, 0, 0, 0, 0, 1, 0, 1, 0, 0];
  const startData2 = const [0x02]; // two bit pixels, indexes are 0-3.
  const clear2 = "100"; // clear code for two bit data
  const end2 = "101"; // end code for two bit data
  const trailer = const [0x3b];

  int bits(String s) => int.parse(s, radix: 2);

  test('one black pixel', () {
    final List<int> bytes = gif.makeGif(1, 1, [0, 0, 0, 0]);

    final expected = <int>[]
      ..addAll(headerBlock)
      ..addAll(oneByOneTwoColor)
      ..addAll(blackBlack)
      ..addAll(startOneByOneImage)
      ..addAll(startData2)
      ..addAll([2, bits("01000$clear2"), bits("1"), 0x00])
      ..addAll(trailer);

    expect(bytes, expected);
  });

  const oneByTwoTwoColor = const [
    0x1, 0, 0x2, 0, 0xF0, 0, 0 // 1x2, 2 colors, background index 0
  ];
  const startOneByTwoImage = const [0x2c, 0, 0, 0, 0, 1, 0, 2, 0, 0];

  test('two black pixels', () {
    final List<int> bytes = gif.makeGif(1, 2, [0, 0, 0, 0, 0, 0, 0, 0]);

    final expected = <int>[]
      ..addAll(headerBlock)
      ..addAll(oneByTwoTwoColor)
      ..addAll(blackBlack)
      ..addAll(startOneByTwoImage)
      ..addAll(startData2)
      ..addAll([2, bits("00000$clear2"), bits("${end2}0"), 0x00])
      ..addAll(trailer);

    expect(bytes, expected);
  });

  const oneByThreeTwoColor = const [
    0x1, 0, 0x3, 0, 0xF0, 0, 0 // 1x3, 2 colors, background index 0
  ];
  const startOneByThreeImage = const [0x2c, 0, 0, 0, 0, 1, 0, 3, 0, 0];

  test('three black pixels', () {
    final bytes = gif.makeGif(1, 3, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

    final expected = <int>[]
      ..addAll(headerBlock)
      ..addAll(oneByThreeTwoColor)
      ..addAll(blackBlack)
      ..addAll(startOneByThreeImage)
      ..addAll(startData2)
      ..addAll([2, bits("10000$clear2"), bits("${end2}1"), 0x00])
      ..addAll(trailer);

    expect(bytes, expected);
  });

  const oneByFourTwoColor = const [
    0x1, 0, 0x4, 0, 0xF0, 0, 0 // 1x4, 2 colors, background index 0
  ];
  const startOneByFourImage = const [0x2c, 0, 0, 0, 0, 1, 0, 4, 0, 0];

  test('four black pixels', () {
    final List<int> bytes =
        gif.makeGif(1, 4, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);

    final expected = <int>[]
      ..addAll(headerBlock)
      ..addAll(oneByFourTwoColor)
      ..addAll(blackBlack)
      ..addAll(startOneByFourImage)
      ..addAll(startData2)
      ..addAll([2, bits("10000$clear2"), bits("${end2}0001"), 0x00])
      ..addAll(trailer);

    expect(bytes, expected);
  });

  const oneBySevenTwoColor = const [
    0x1, 0, 0x7, 0, 0xF0, 0, 0 // 1x7, 2 colors, background index 0
  ];
  const startOneBySevenImage = const [0x2c, 0, 0, 0, 0, 1, 0, 7, 0, 0];

  test('seven black pixels', () {
    final List<int> bytes =
        gif.makeGif(1, 7, new List<int>.generate(28, (_) => 0));

    // bumps to 4-bit codes at the last pixel
    final expected = <int>[]
      ..addAll(headerBlock)
      ..addAll(oneBySevenTwoColor)
      ..addAll(blackBlack)
      ..addAll(startOneBySevenImage)
      ..addAll(startData2)
      ..addAll([3, bits("10000$clear2"), bits("00001111"), bits("0101"), 0x00])
      ..addAll(trailer);

    expect(bytes, expected);
  });

  const threeByOneFourColor = const [
    0x3, 0, 0x1, 0, 0xF1, 0, 0 // 1x2, 4 colors, background index 0
  ];
  const startThreeByOneImage = const [0x2c, 0, 0, 0, 0, 3, 0, 1, 0, 0];
  const redGreenBlueBlack = const [
    0xff, 0, 0, 0, 0xff, 0, 0, 0, 0xff, 0, 0, 0 // 3 bytes per color
  ];

  test('rgb', () {
    final List<int> bytes =
        gif.makeGif(3, 1, [0xff, 0, 0, 0, 0, 0xff, 0, 0, 0, 0, 0xff, 0]);

    final expected = <int>[]
      ..addAll(headerBlock)
      ..addAll(threeByOneFourColor)
      ..addAll(redGreenBlueBlack)
      ..addAll(startThreeByOneImage)
      ..addAll(startData2)
      ..addAll([2, bits("01000$clear2"), bits("1010100"), 0x00])
      ..addAll(trailer);

    expect(bytes, expected);
  });
}


test/gif_web_test.dart
@TestOn("browser")

library gif_web_test;

import 'dart:html';

import "package:gifencoder/gifencoder.dart" as gif;
import 'package:test/test.dart';

void main() {
  test('one black pixel', () {
    final ctx = new CanvasElement(width: 1, height: 1).context2D
      ..fillStyle = "black"
      ..fillRect(0, 0, 1, 1);
    final data = ctx.getImageData(0, 0, 1, 1);
    final blob = new Blob([gif.makeGif(data.width, data.height, data.data)]);

    final f = new FileReader();

    void loadedDataUrl(ProgressEvent e) {
      expect(f.error, null);
      expect(f.readyState, FileReader.DONE);
      String url = f.result;
      expect(url, startsWith("data:;"));
      url = url.replaceFirst("data:;", "data:image/gif;");
      final img = new ImageElement(src: url);
      img.onLoad.listen(expectAsync1((e) {
        expect(img.width, 1);
        expect(img.height, 1);
        checkPixel(img, 0, 0, [0, 0, 0, 255]);
      }));
    }

    f.onLoadEnd.listen(expectAsync1(loadedDataUrl));
    f.readAsDataUrl(blob);
  });
}

void checkPixel(ImageElement img, int x, int y, List<int> expectedRGBA) {
  final ctx = new CanvasElement(width: x + 1, height: y + 1).context2D
    ..fillStyle = "white"
    ..fillRect(0, 0, x + 1, y + 1)
    ..drawImage(img, 0, 0);
  final data = ctx.getImageData(0, 0, 1, 1).data;
  expect(data, expectedRGBA);
}


lib/src/lzw.dart
library lzw;

/// Compresses pixels using LZW.
List<int> compress(List<int> pixels, int colorBits) {
  final book = new CodeBook(colorBits);
  final buf = new CodeBuffer(book)..add(book.clearCode);

  if (pixels.isEmpty) {
    buf.add(book.endCode);
    return buf.finish();
  }

  int code = pixels[0];
  for (int px in pixels.sublist(1)) {
    final int newCode = book.codeAfterAppend(code, px);
    if (newCode == null) {
      buf.add(code);
      book.define(code, px);
      code = px;
    } else {
      code = newCode;
    }
  }
  buf..add(code)..add(book.endCode);
  return buf.finish();
}

/// The highest code that can be defined in the CodeBook.
const int maxCode = (1 << 12) - 1;

/// A CodeBook contains codes defined during LZW compression. It's a mapping from a string
/// of pixels to the code that represents it. The codes are stored in a trie which is
/// represented as a map. Codes may be up to 12 bits. The size of the codebook is always
/// the minimum power of 2 needed to represent all the codes and automatically increases
/// as new codes are defined.
class CodeBook {
  int colorBits;

  ///  The "clear" code which resets the table.
  int clearCode;

  ///  The "end of data" code.
  int endCode;

  // A mapping from (c1, pixel) -> c2 that returns the new code for the pixel string
  // formed by appending a pixel to the end of c1's pixel string. (In addition, the
  // codes for single pixels are stored in the map with c1 set to 0.)
  // The key is encoded by shifting c1 to the left by eight bits and adding the pixel,
  // forming a 20-bit number.
  Map<int, int> _codeAfterAppend;

  /// Codes from this value and above are not yet defined.
  int nextUnused;

  ///  The number of bits required to represent every code.
  int bitsPerCode;

  /// The current size of the codebook.
  int size;

  CodeBook(this.colorBits) {
    if (colorBits < 2) {
      colorBits = 2;
    }
    assert(colorBits <= 8);
    clearCode = 1 << colorBits;
    endCode = clearCode + 1;
    clear();
  }

  void clear() {
    _codeAfterAppend = <int, int>{};
    nextUnused = endCode + 1;
    bitsPerCode = colorBits + 1;
    size = 1 << bitsPerCode;
  }

  /// Returns the new code after appending a pixel to the pixel string represented by the previous code,
  /// or null if the code isn't in the table.
  int codeAfterAppend(int code, int pixelIndex) =>
      _codeAfterAppend[(code << 8) | pixelIndex];

  /// Defines a new code to be the pixel string of a previous code with one pixel appended.
  /// Returns true if defined, or false if there's no more room in the table.
  bool define(int code, int pixelIndex) {
    if (nextUnused == maxCode) {
      return false;
    }
    _codeAfterAppend[(code << 8) | pixelIndex] = nextUnused++;
    if (nextUnused > size) {
      bitsPerCode++;
      size = 1 << bitsPerCode;
    }
    return true;
  }
}

/// Writes a sequence of integers using a variable number of bits, for LZW compression.
class CodeBuffer {
  final CodeBook book;
  final List<int> finishedBytes = <int>[];

  /// A buffer containing bits not yet added to finishedBytes.
  int buf = 0;

  /// Number of bits in the buffer.
  int bits = 0;

  CodeBuffer(this.book);

  void add(int code) {
    assert(code >= 0 && code < book.size);
    buf |= (code << bits);
    bits += book.bitsPerCode;
    while (bits >= 8) {
      finishedBytes.add(buf & 0xFF);
      buf = buf >> 8;
      bits -= 8;
    }
  }

  List<int> finish() {
    // Add the remaining bits. (Unused bits are set to zero.)
    if (bits > 0) {
      finishedBytes.add(buf);
    }

    // The final result starts withe the number of color bits.
    final dest = <int>[]..add(book.colorBits);

    // Divide it up into blocks with a size in front of each block.
    for (int i = 0, len = finishedBytes.length; i < len;) {
      if (len - i >= 255) {
        dest
          ..add(255)
          ..addAll(finishedBytes.sublist(i, i + 255));
        i += 255;
      } else {
        dest
          ..add(len - i)
          ..addAll(finishedBytes.sublist(i, len));
        i = len;
      }
    }
    dest.add(0);
    return dest;
  }
}


example/example.dart
import 'dart:async';
import 'dart:html';
import 'dart:typed_data';

import "package:gifencoder/gifencoder.dart" as gifencoder;

const int size = 200;
const int frameCount = 32;
const int framesPerSecond = 15;

void drawSquare(CanvasRenderingContext2D ctx, int frameNumber) {
  final frameFraction = (frameNumber / frameCount);
  final green = 0;
  final blue = 0;
  for (int i = 0; i < size / 2; i++) {
    final drawnFraction = i / (size / 2);
    final red = (256 * (frameFraction + drawnFraction)).floor() % 256;
    ctx
      ..fillStyle = "rgb($red,$green,$blue)"
      ..fillRect(i, i, size - i * 2, size - i * 2);
  }
}

Future<String> createDataUrl(Uint8List bytes) {
  final c = new Completer<String>();
  final f = new FileReader();
  f.onLoadEnd.listen((ProgressEvent e) {
    if (f.readyState == FileReader.DONE) {
      final String url = f.result;
      c.complete(url.replaceFirst("data:;", "data:image/gif;"));
    }
  });
  f.readAsDataUrl(new Blob([bytes]));
  return c.future;
}

void main() {
  final ctx = new CanvasElement(width: size, height: size).context2D;
  final frames = new gifencoder.GifBuffer(size, size);
  for (var i = 0; i < frameCount; i++) {
    drawSquare(ctx, i);
    frames.add(ctx.getImageData(0, 0, size, size).data);
  }
  final Uint8List gif = frames.build(framesPerSecond);
  createDataUrl(gif).then((dataUrl) {
    (querySelector("#gif") as ImageElement)..src = dataUrl;
  });
}


lib/src/gifencoder.dart
import "dart:typed_data";
import "lzw.dart" as lzw;

// Spec: http://www.w3.org/Graphics/GIF/spec-gif89a.txt
// Explanation: http://www.matthewflickinger.com/lab/whatsinagif/bits_and_bytes.asp
// Also see: http://en.wikipedia.org/wiki/File:Quilt_design_as_46x46_uncompressed_GIF.gif

const int maxColorBits = 8;
const int maxColors = 1 << maxColorBits;

/// Creates a GIF from per-pixel rgba data, ignoring the alpha channel.
/// Returns a list of bytes. Throws an exception if the the image has too
/// many colors.
///
/// (The input format is the same as the "data" field of the html.ImageData class,
/// which can be created from a canvas element.)
Uint8List makeGif(int width, int height, List<int> rgba) {
  final b = new GifBuffer(width, height)..add(rgba);
  return b.build(1);
}

///  An incomplete GIF, possibly animated.
class GifBuffer {
  final int width;
  final int height;
  final _colors = new _ColorTableBuilder();
  final _frames = <Uint8List>[];

  /// Creates an incomplete gif of the specified width and height and zero frames.
  GifBuffer(this.width, this.height);

  /// Adds a frame to the animation. The pixels are specified as rgba data but the alpha channel is
  /// ignored. Throws an exception if we run out of colors.
  void add(List<int> rgba) {
    _frames.add(_colors.indexImage(width, height, rgba));
  }

  /// Returns the bytes of the GIF. If more than one frame has been added, it will be animated.
  Uint8List build(int framesPerSecond) {
    final colors = _colors.build();
    int delay = 100 ~/ framesPerSecond;
    if (delay < 6) {
      // http://nullsleep.tumblr.com/post/16524517190/animated-gif-minimum-frame-delay-browser-compatibility
      delay = 6;
    }

    final bytes = <List<int>>[]
      ..add(_header(width, height, colors.bits))
      ..add(colors.table);

    if (_frames.length <= 1) {
      // not animated
      if (_frames.length == 1) {
        bytes
          ..add(_startImage(0, 0, width, height))
          ..add(lzw.compress(_frames[0], colors.bits));
      }
    } else {
      bytes.add(_loop(0));

      for (int i = 0; i < _frames.length; i++) {
        final frame = _frames[i];
        bytes
          ..add(_delayNext(delay))
          ..add(_startImage(0, 0, width, height))
          ..add(lzw.compress(frame, colors.bits));
      }
    }
    bytes.add(_trailer());

    int len = 0;
    for (var chunk in bytes) {
      len += chunk.length;
    }

    final result = new Uint8List(len);
    int offset = 0;
    for (var chunk in bytes) {
      result.setRange(offset, offset + chunk.length, chunk);
      offset += chunk.length;
    }
    assert(offset == len);
    return result;
  }
}

class _ColorTableBuilder {
  final List<int> table = <int>[];
  final Map<int, int> colorToIndex = <int, int>{};
  int bits;

  /// Given rgba data, add each color to the color table.
  /// Returns the same pixels as color indexes.
  /// Throws an exception if we run out of colors.
  Uint8List indexImage(int width, int height, List<int> rgba) {
    final pixels = new Uint8List(width * height);
    assert(pixels.length == rgba.length / 4);
    for (int i = 0; i < rgba.length; i += 4) {
      final int color = rgba[i] << 16 | rgba[i + 1] << 8 | rgba[i + 2];
      int index = colorToIndex[color];
      if (index == null) {
        if (colorToIndex.length == maxColors) {
          throw new Exception("image has more than $maxColors colors");
        }
        index = table.length ~/ 3;
        colorToIndex[color] = index;
        table..add(rgba[i])..add(rgba[i + 1])..add(rgba[i + 2]);
      }
      pixels[i >> 2] = index;
    }
    return pixels;
  }

  /// Pads the color table with zeros to the next power of 2 and sets bits.
  _ColorTable build() {
    for (int bits = 1; bits <= 8; bits++) {
      final int colors = 1 << bits;
      if (colors * 3 >= table.length) {
        final copy = new Uint8List(colors * 3)
          ..setRange(0, table.length, table);
        return new _ColorTable(bits, copy);
      }
    }
    throw new Exception("internal error; too many colors");
  }
}

class _ColorTable {
  final int bits;
  final Uint8List table;

  _ColorTable(this.bits, this.table);

  int get numColors => table.length ~/ 3;
}

List<int> _header(int width, int height, int colorBits) {
  const _headerBlock = const [0x47, 0x49, 0x46, 0x38, 0x39, 0x61]; // GIF 89a

  final bytes = <int>[]..addAll(_headerBlock);
  _addShort(bytes, width);
  _addShort(bytes, height);
  bytes..add(0xF0 | colorBits - 1)..add(0)..add(0);
  return bytes;
}

// See: http://odur.let.rug.nl/~kleiweg/gif/netscape.html
List<int> _loop(int reps) {
  final bytes = <int>[0x21, 0xff, 0x0B]
    ..addAll("NETSCAPE2.0".codeUnits)
    ..addAll([3, 1]);
  _addShort(bytes, reps);
  bytes.add(0);
  return bytes;
}

List<int> _delayNext(int centiseconds) {
  final bytes = <int>[0x21, 0xF9, 4, 0];
  _addShort(bytes, centiseconds);
  bytes..add(0)..add(0);
  return bytes;
}

List<int> _startImage(int left, int top, int width, int height) {
  final bytes = <int>[0x2C];
  _addShort(bytes, left);
  _addShort(bytes, top);
  _addShort(bytes, width);
  _addShort(bytes, height);
  bytes.add(0);
  return bytes;
}

List<int> _trailer() => [0x3b];

void _addShort(List<int> dest, int n) {
  if (n < 0 || n > 0xFFFF) throw new Exception("out of range for short: $n");
  dest..add(n & 0xff)..add(n >> 8);
}
